packages/components/src/presentation-components/properties/inputs/ItemsLoader.tsx

```diff
@@ -0,0 +1,97 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+/** @internal */
+export const VALUE_BATCH_SIZE = 100;
+
+/** @internal */
+export const FILTER_WARNING_OPTION = { label: "Too many values please use filter", value: "__filter__", disabled: true };
+
+interface LoadedItems<T> {
+  options: T[];
+  length: number;
+  hasMore: boolean;
+}
+
+/** @internal */
+export class ItemsLoader<T> {
+  private _loadedItems: T[] = [];
+  private _isLoading = false;
+  private _hasMore = true;
+  private _disposed = false;
+  private _offset = 0;
+
+  constructor(
+    private _beforeLoad: () => void,
+    private _onItemsLoaded: (newItems: T[]) => void,
+    private _loadItems: (offSet: number) => Promise<LoadedItems<T>>,
+    private _getOptionLabel: (option: T) => string,
+  ) {}
+
+  public dispose() {
+    this._disposed = true;
+  }
+
+  public async loadInitialItems(initialSelectedValues?: string[]) {
+    const needsItemsLoaded = (items: T[]) => {
+      if (initialSelectedValues && initialSelectedValues.length > 0) {
+        const matchingItems = items.filter((item) => {
+          return initialSelectedValues.some((initialSelectedValue) => initialSelectedValue === this._getOptionLabel(item));
+        });
+        return matchingItems.length < initialSelectedValues.length;
+      }
+      return items.length < VALUE_BATCH_SIZE;
+    };
+
+    await this.loadUniqueItems(needsItemsLoaded);
+  }
+
+  public async loadItems(filterText?: string) {
+    if (!filterText) {
```

Comment:
I would move this into `needsItemsLoaded` to make sure that if `filterText` is not specified and no items are loaded we would load atleast one batch of items.

Reply:
changed